crates/net/network/src/network.rs

```diff
@@ -247,18 +248,38 @@ impl NetworkInfo for NetworkHandle {
     fn is_syncing(&self) -> bool {
         SyncStateProvider::is_syncing(self)
     }
+
+    fn is_initially_syncing(&self) -> bool {
+        SyncStateProvider::is_initially_syncing(self)
+    }
 }
 
 impl SyncStateProvider for NetworkHandle {
     fn is_syncing(&self) -> bool {
         self.inner.is_syncing.load(Ordering::Relaxed)
     }
+    // used to guard the txpool
+    fn is_initially_syncing(&self) -> bool {
+        let init_sync_done = self.inner.initial_sync_done.load(Ordering::Relaxed);
+        let is_currently_syncing = self.inner.is_syncing.load(Ordering::Relaxed);
+        return is_currently_syncing && !init_sync_done
+    }
 }
 
 impl NetworkSyncUpdater for NetworkHandle {
     fn update_sync_state(&self, state: SyncState) {
-        let is_syncing = state.is_syncing();
-        self.inner.is_syncing.store(is_syncing, Ordering::Relaxed)
+        let future_state = state.is_syncing();
+        let curr_state = self.inner.is_syncing.load(Ordering::Relaxed);
+        let init_sync_done = self.inner.initial_sync_done.load(Ordering::Relaxed);
+        self.inner.is_syncing.store(future_state, Ordering::Relaxed);
+        if !init_sync_done {
+            // we've either been in Idle or pipeline Syncing here, check to see if we're
+            // moving from a pipeline sync to idle
+            if curr_state && !future_state {
+                // set initial_sync_done flag to true here, every sync after this will be a livesync
+                self.inner.initial_sync_done.store(true, Ordering::Relaxed);
+            }
+        }
```

Comment:
I think we can simplify this,
basically `initial_sync_done` will be true when `is_syncing` goes from false to true

I think we can use `.swap` here and compare the replaced value

Reply:
mm let me think about this - I think I see what you're saying, but I need think through the implementation

I think you're right in that we know when `is_syncing` goes from false-> true (the first time), `initial_sync_done` -> true

Reply:
yeah I think this will replace the block that I have now:
```
self.inner.initial_sync_done.compare_exchange(false, curr_state && !future_state, Ordering::Relaxed, Ordering::Relaxed);
```

basically a CAS (`compare_and_swap` seems deprecated in favor of this new `compare_exchange` api)  that only sets `initial_sync_done` if its false

I do want to mention that `initial_sync_done` should only be true when `!init_sync_done` && `is_syncing` goes from true->false

what you mentioned may not work (when `is_syncing` goes from false->true) because I think we need to capture when the initial_sync completes and not just when it starts.


otherwise in the network manager code we won't be able to differentiate between a live_sync and an ongoing pipeline sync (i.e is_syncing() && is_initial_syncing()` will be true even during the pipeline sync

Reply:
what we want to capture is the state transition from false->true->false with `is_syncing`

false->true show cases that we've kicked off a pipeline sync, but we can't mark `is_initial_done` until it has actually completed